docs/development-guide/extending-gaffer.md

```diff
@@ -0,0 +1,218 @@
+This class may appear very simple but the real functionality for this generator
+is in the `OpenCypherCsvElementGenerator` that it extends. This generator is an
+abstract class that implements the `ElementGenerator` interface and has a lot of
+functionality for handling Open Cypher CSVs.
+
+You can view the [full class in the repository](https://github.com/gchq/Gaffer/blob/develop/core/data/src/main/java/uk/gov/gchq/gaffer/data/generator/OpenCypherCsvElementGenerator.java)
+which has the logic for converting the CSV format but the key bit to take note
+of is the `apply(final Iterable<? extends String> strings)` method. This method
+is the main override which will be passed the lines of a CSV file to convert and
+return a `Iterable<? extends Element>` e.g. a list of `Elements`.
+
+This pattern is repeated across many of the different generators and can be used
+as inspiration for your custom generator. As with the Operations so long as
+you're implementing the appropriate interfaces it's hard to go far wrong.
+
+### Writing an Aggregation Function
+
+We can provide a custom aggregation function in a number of ways, most of the
+aggregation functionality is provided by the Koryphe library sat adjacent to
+Gaffer. If we want to provide a new "generic" aggregator we would add it in this
+Koryphe library. For example lets take a look at the very simple `Max` comparator,
+this takes a pair of Java 8 `Comparables` and finds the highest value one, this
+function is applied as an aggregation.
+
+```java
+@Since("1.0.0")
+@Summary("Calculates the max value")
+public class Max extends KorypheBinaryOperator<Comparable> {
+    @Override
+    protected Comparable _apply(final Comparable a, final Comparable b) {
+        return a.compareTo(b) >= 0 ? a : b;
+    }
+}
+```
+
+If we want to add something very specific to a store type or some other
+restriction we can add this into the appropriate Store location within Gaffer.
+An example of this is the `HyperLogLogPlusAggregator` in the Sketches library,
+this merges HLLPs together.
+
+```java
+@Since("1.0.0")
+@Summary("Aggregates HyperLogLogPlus objects")
+@Deprecated
+public class HyperLogLogPlusAggregator extends KorypheBinaryOperator<HyperLogLogPlus> {
+    @Override
+    protected HyperLogLogPlus _apply(final HyperLogLogPlus a, final HyperLogLogPlus b) {
+        try {
+            a.addAll(b);
+        } catch (final CardinalityMergeException exception) {
+            throw new RuntimeException("An Exception occurred when trying to aggregate the HyperLogLogPlus objects", exception);
+        }
+        return a;
+    }
+}
+```
+
+You can also add aggregators by loading a JAR into Accumulo and
+accessing it on the classpath, this is more something an end user customising
+their usage of the platform might do rather than a developer. For development
+purposes it's preferred to go with the hierarchy of does it belong in Koryphe?
+If no, is it store specific or can live inside a library? If no then it could be
```

Comment:
I'm not sure about this. I think the point being made is that some functions might be better placed in Koryphe. This doesn't make much sense for an external developer who probably isn't looking to make contributions to Koryphe.